prusti-interface/src/specs/checker.rs

```diff
@@ -0,0 +1,160 @@
+use log::{debug, info};
+use rustc_hir::{
+    self as hir,
+    def_id::{DefId, LocalDefId},
+    intravisit::{self, Visitor},
+    itemlikevisit::ItemLikeVisitor,
+};
+use rustc_middle::{hir::map::Map, ty::TyCtxt};
+use rustc_span::{MultiSpan, Span};
+
+use std::collections::HashMap;
+
+use crate::{
+    environment::Environment,
+    utils::{has_prusti_attr, has_spec_only_attr},
+    PrustiError,
+};
+
+/// Checker visitor for the specifications. Currently checks that `#[predicate]`
+/// functions are never called from non-specification code, but more checks may follow.
+pub struct SpecChecker {
+    /// Map of the `DefID`s to the `Span`s of `#[predicate]` functions found in the first pass.
+    predicates: HashMap<DefId, Span>,
+
+    /// Span of call and definition of predicates illegally called, collected in the second pass.
+    pred_calls: Vec<(Span, Span)>,
```

Comment:
`pred_usages` would be a better name, because of 0bab6b182dbfdf47bec7292b9a83566a0f20e468. Similarly, the comment could say "usage" instead of "call".

Reply:
updated in d4e946de.

i switched to many separate commits, mainly because i hope it makes reviewing easier; at least it helps me keep an overview. if you want, we can switch back though, e.g. for nicer git-history or such.